rasa/utils/train_utils.py

```diff
@@ -335,3 +342,97 @@ def override_defaults(
                 config[key] = custom[key]
 
     return config
+
+
+def create_data_generators(
+    model_data: RasaModelData,
+    batch_sizes: Union[int, List[int]],
+    epochs: int,
+    batch_strategy: Text = SEQUENCE,
+    eval_num_examples: int = 0,
+    random_seed: Optional[int] = None,
+) -> Tuple[RasaBatchDataGenerator, Optional[RasaBatchDataGenerator]]:
+    """Create data generators for train and optional validation data.
+
+    Args:
+        model_data: The model data to use.
+        batch_sizes: The batch size(s).
+        epochs: The number of epochs to train.
+        batch_strategy: The batch strategy to use.
+        eval_num_examples: Number of examples to use for validation data.
+        random_seed: The random seed.
+
+    Returns:
+        The training data generator and optional validation data generator.
+    """
+    validation_data_generator = None
+    if eval_num_examples > 0:
+        model_data, evaluation_model_data = model_data.split(
+            eval_num_examples, random_seed,
+        )
+        validation_data_generator = RasaBatchDataGenerator(
+            evaluation_model_data,
+            batch_size=batch_sizes,
+            epochs=epochs,
+            batch_strategy=batch_strategy,
+            shuffle=True,
+        )
+
+    data_generator = RasaBatchDataGenerator(
+        model_data,
+        batch_size=batch_sizes,
+        epochs=epochs,
+        batch_strategy=batch_strategy,
+        shuffle=True,
+    )
+
+    return data_generator, validation_data_generator
```

Comment:
Sort of related to the comment below (so it's better to read the one below first).

In this case it might also be a good idea to create two separate factory functions like `RasaBatchDataGenerator.new_test_generator` and `RasaBatchDataGenerator.new_generator` instead — just for the sake of having those functions in the right place (and not in `utils`) and making them easier to read + easier to test.

Reply:
Would not create factory methods for this. `RasaBatchDataGenerator.new_test_generator` and `RasaBatchDataGenerator.new_generator` would be the same method, the same as the constructor, as I would not put the logic 
```
    if eval_num_examples > 0:
        model_data, evaluation_model_data = model_data.split(
            eval_num_examples, random_seed,
        )
```
inside the `RasaBatchDataGenerator`. This logic rather belongs to the model itself. I just extracted it to a separate function as we use this logic across all our custom models.

Reply:
Should I move the code to the model itself again?

I think once we refactor the whole structure around ML components, this can be improved further. But that is something for later.

Reply:
I think it can certainly become a factory method inside `RasaBatchDataGenerator` but if you say that the refactoring is going to continue further then it might be fine to skip it for now as well.

Reply:
same argument here as with the other function. it is used in both TED and DIET. creating validation dataset is external logic towards `RasaBatchDataGenerator`